archunit/src/main/java/com/tngtech/archunit/library/freeze/FreezingArchRule.java

```diff
@@ -0,0 +1,219 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.tngtech.archunit.library.freeze;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+
+import com.tngtech.archunit.ArchConfiguration;
+import com.tngtech.archunit.PublicAPI;
+import com.tngtech.archunit.base.Predicate;
+import com.tngtech.archunit.core.domain.JavaClasses;
+import com.tngtech.archunit.lang.ArchRule;
+import com.tngtech.archunit.lang.EvaluationResult;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.tngtech.archunit.PublicAPI.Usage.ACCESS;
+import static com.tngtech.archunit.library.freeze.ViolationStoreFactory.FREEZE_STORE_PROPERTY;
+
+/**
+ * A decorator around an existing {@link ArchRule} that "freezes" the state of all violations on the first call instead of failing the test.
+ * This means in particular that the first run of a {@link FreezingArchRule} will always pass, consecutive calls will only fail if "unknown"
+ * violations are introduced (read below for further explanations when a violation is "unknown").
+ * <br><br>
+ * You might consider using this class, if you introduce a new {@link ArchRule} to an existing project that causes too many violations to solve
+ * at the current time. A typical example is a huge legacy project where a new rule might cause thousands of violations. Even if it is impossible
+ * to fix all those violations at the moment, it is typically a good idea to a) make sure no further violations are introduced and
+ * b) incrementally fix those violations over time one by one.
+ * <br><br>
+ * {@link FreezingArchRule} uses two concepts to support this use case. First a {@link ViolationStore} to store the result of the current
+ * evaluation of this rule (compare {@link #persistIn(ViolationStore)}). Second a {@link ViolationLineMatcher} to decide which violations are "known",
+ * i.e. have been present from the beginning (compare {@link #associateViolationLinesVia(ViolationLineMatcher)}).
+ * The reason to adjust the {@link ViolationLineMatcher} and not simply check for equality might be to make the comparison for known violations
+ * more resilient, e.g. by ignoring the current line number (assume a class has 500 lines, adding a line at the beginning should maybe not affect an
+ * unrelated known violation in line 490).
+ * <br><br>
+ * If you do not configure {@link ViolationStore} or {@link ViolationLineMatcher}, a default will be used (compare the javadoc of the
```

Comment:
Yeah, you seem right, odd :frowning: I'm pretty sure I have described that at some point, but I can't find it either. Maybe it was just a draft. But yes, I don't think the default should be documented at the interfaces.
I would document it at `FreezingArchRule`, because in the end that class decides which default to use. What do you think? I think it's important to document the defaults within the Javadoc.

Reply:
I absolutely agree. If you look at [my fork](https://github.com/hankem/ArchUnit/commits/freeze_rules), you'll already find a suggestion for that. 😉

Reply:
(If you prefer, I could create a PR for my tiny review commits to be added on top of your entire PR.)

Reply:
Yes, feel free to create a PR. I would include the configuration of `ViolationStore` and `ViolationLineMatcher` via `archunit.properties` in your adjusted Javadoc of `FreezingArchRule` as well. What do you think? I think this is currently only documented in the user guide, but I think it should also go into the Javadoc.
I.e. that you can use `persistIn(store)`, but also can configure it via `archunit.properties`. And likewise for `associateViolationLinesVia(matcher)`.

Reply:
#189 it is.

Reply:
I think that documenting the configurability via archunit.properties on the JavaDoc is a very good idea!
I wonder whether one could actually move some JavaDoc from FreezingArchRule to the two factories. They are, admittedly, only